# Data cleaning using pandas

In this notebook, you'll see how clean structured data using `pandas`. We are aware that you have already done an extensive review of pandas, and are now experts in data manipulation, so we'll keep it short.

We are going to use a dataset about New York City Squirrels taken from [NYC Open Data](https://opendata.cityofnewyork.us/) which is a great resource for data sets.
This dataset is quite small and this notebook will just be an overview of different stuff to look out for when cleaning data.


In [2]:
import pandas as pd
import numpy as np

In [3]:
filename = "nyc_squirrels.csv"
df = pd.read_csv(filename)
df.head()

X          Y Unique Squirrel ID Hectare Shift      Date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   Hectare Squirrel Number    Age Primary Fur Color Highlight Fur Color  ...  \
0                        3    NaN               NaN                 NaN  ...   
1                        4    NaN               NaN                 NaN  ...   
2                        8    NaN              Gray                 NaN  ...   
3                       14  Adult              Gray                 NaN  ...   
4                        5  Adult              Gray            Cinnamon  ...   

    Kuks  Quaas  Moans Tail flags Tail twitches  Approaches  Indifferent  \
0  False  False  False      False         False       False        False   
1  False  False  False      False         False       False        False   
2  False  False  False      False         False       False        False   
3  False  False  False      False         False       False        False   
4  False  False  False      False         False       False        False   

   Runs from  Other Interactions                                    Lat/Long  
0      False                 NaN  POINT (-73.9561344937861 40.7940823884086)  
1      False                 NaN  POINT (-73.9688574691102 40.7837825208444)  
2      False                 NaN  POINT (-73.97428114848522 40.775533619083)  
3       True                 NaN  POINT (-73.9596413903948 40.7903128889029)  
4      False                 NaN  POINT (-73.9702676472613 40.7762126854894)  

[5 rows x 31 columns]

## Things to check

### Missing values

There are different ways that a value can be missing. It can be just an empty string (`""`), or `NaN` (Not a Number).
`np.nan`, `None` and `NaT` (for `datetime64` types) are standard missing value for Pandas.

You can either delete the rows/attribute if there are too many missing, or replace them. I can't give you a certain threshold to delete or replace, it really depends on the variable, on the context, and what you are going to do with the data later on.
In the same vein, there are many different ways to replace a value, and it will depend on the type of data you have. [[More info here]](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html)

Pandas provides `isnull()`, `isna()` functions to detect missing values. Both of them do the same thing.

In [4]:
print("There are {} rows of data".format(len(df)))

There are 3023 rows of data


We see that we have 3023 rows of data. Looking at the number of NaN values per column, one can confidently say that columns where we have > 1000 missing values are not worth keeping. The case can be made that, for certain scenarios, the columns `Highlight Fur Color` should be kept too, but here we are going to drop it because the column `Combination of Primary and Highlight Color` combines both.

This notebook is going to feature small code blocks that you'll have to complete.

The first task is to count how many nan values there are per attribute. Hint: use `df.isnull()`

In [5]:
# Check how many rows of each attribute are NaN
total_rows = format(len(df))

print("COLUMNS TO DELETE : ")
print("-------------------")
for column in df.columns:
    if df[column].isnull().sum() > 1000:
        print(f"{column} ({df[column].isnull().sum()}/{total_rows} missing values)")

COLUMNS TO DELETE : 
-------------------
Highlight Fur Color (1086/3023 missing values)
Color notes (2841/3023 missing values)
Specific Location (2547/3023 missing values)
Other Activities (2586/3023 missing values)
Other Interactions (2783/3023 missing values)


Now that we've identified all the columns that have NaN values, we would like to drop those with more than 1000 NaN values.

Can you do that? Hint: `df.drop()`
  
Moreover, can you print the columns that were dropped, along with the number of missing values that they had?

In [6]:
# Drop the columns that have more than 1000 missing values
for column in df.columns:
    if df[column].isnull().sum() > 1000:
        print(f"\"{column}\" will be dropped ({df[column].isnull().sum()}/{total_rows} missing values)")
        df.drop(columns=column, inplace=True) #inplace=True makes sure that we drop it in df and not create a new one
        
print(df.head())        
# Print the columns that were dropped

"Highlight Fur Color" will be dropped (1086/3023 missing values)
"Color notes" will be dropped (2841/3023 missing values)
"Specific Location" will be dropped (2547/3023 missing values)
"Other Activities" will be dropped (2586/3023 missing values)
"Other Interactions" will be dropped (2783/3023 missing values)
           X          Y Unique Squirrel ID Hectare Shift      Date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   Hectare Squirrel Number    Age Primary Fur Color  \
0                        3    NaN               NaN   
1                        4    NaN               NaN   
2                        8    NaN              Gray   
3                       14  Adult              G

### Attribute type

Sometimes, an attribute does not have the correct type, e.g. integers are labeled as floats. So it is useful to check if it needs to be changed or not.


Please print the dtypes of every attribute.

In [7]:
# Print the dtype of every attribute
df.dtypes

X                                             float64
Y                                             float64
Unique Squirrel ID                             object
Hectare                                        object
Shift                                          object
Date                                            int64
Hectare Squirrel Number                         int64
Age                                            object
Primary Fur Color                              object
Combination of Primary and Highlight Color     object
Location                                       object
Above Ground Sighter Measurement               object
Running                                          bool
Chasing                                          bool
Climbing                                         bool
Eating                                           bool
Foraging                                         bool
Kuks                                             bool
Quaas                       

For example, for `Hectare`, one could think that the variable is just the size of the plot of land. However, it seems it is just an identifier for which plot of land it is. One would have to see the documentation to know for sure. In this case, the `object` dtype is appropriate.

In [8]:
# Print unique values of the "Hectare" attribute
df["Hectare"].drop_duplicates()

0       37F
1       21B
2       11B
3       32E
4       13E
       ... 
2415    10E
2447    25B
2654    20E
2801    38H
2873    30F
Name: Hectare, Length: 339, dtype: object

### Attribute values

Carefully check each attribute to see if the unique values are coherent and make sense.
Moreover, check that minimum and maximum values make sense. (e.g. age of a person that is negative)

Hint: use your previously acquired knowledge of extracting unique values.
Be careful to limit the attributes that you print by the number of unique values they have, as there are fields that have all their values that are unique, and we don't want a kilometer long list of each value.

In [9]:
# Print the number of unique values of each column, only printing those that have less than 100 unique values
for column in df.columns:
    print("\n")
    print(f"{column} unique values : {df[column].nunique()}")
    # min and max cause issues for Age because there's a FALSE or something
    if df[column].nunique() < 100:
        print(f"----- {column} VALUES -----")
        print(df[column].drop_duplicates())
        



X unique values : 3023


Y unique values : 3023


Unique Squirrel ID unique values : 3018


Hectare unique values : 339


Shift unique values : 2
----- Shift VALUES -----
0    PM
1    AM
Name: Shift, dtype: object


Date unique values : 11
----- Date VALUES -----
0     10142018
1     10192018
3     10172018
5     10102018
7     10082018
8     10062018
10    10132018
12    10072018
15    10182018
55    10122018
69    10202018
Name: Date, dtype: int64


Hectare Squirrel Number unique values : 23
----- Hectare Squirrel Number VALUES -----
0        3
1        4
2        8
3       14
4        5
6        2
8        1
11       6
13       9
16      23
19       7
42      13
64      10
93      16
122     11
172     12
463     20
469     15
521     19
757     17
1322    21
1398    22
2983    18
Name: Hectare Squirrel Number, dtype: int64


Age unique values : 3
----- Age VALUES -----
0            NaN
3          Adult
17      Juvenile
1458           ?
Name: Age, dtype: object


Primary Fur Color

### Notes on unique values that might cause an issue

- "Age" has a "?"
- "Combination of Primary and Highlight Color" has "+" in strings (ok?)
- "Above ground sighter measurement" has mostly numbers but one "FALSE" and one "NaN"


### Pretty useful !

In [10]:
df.describe(include="all").T  # Transpose the data frame so that it fits in a cell

count unique  \
X                                           3023.0    NaN   
Y                                           3023.0    NaN   
Unique Squirrel ID                            3023   3018   
Hectare                                       3023    339   
Shift                                         3023      2   
Date                                        3023.0    NaN   
Hectare Squirrel Number                     3023.0    NaN   
Age                                           2902      3   
Primary Fur Color                             2968      3   
Combination of Primary and Highlight Color    3023     22   
Location                                      2959      2   
Above Ground Sighter Measurement              2909     41   
Running                                       3023      2   
Chasing                                       3023      2   
Climbing                                      3023      2   
Eating                                        3023      2   
Foraging                                      3023      2   
Kuks                                          3023      2   
Quaas                                         3023      2   
Moans                                         3023      2   
Tail flags                                    3023      2   
Tail twitches                                 3023      2   
Approaches                                    3023      2   
Indifferent                                   3023      2   
Runs from                                     3023      2   
Lat/Long                                      3023   3023   

                                                                                   top  \
X                                                                                  NaN   
Y                                                                                  NaN   
Unique Squirrel ID                                                       7D-PM-1010-01   
Hectare                                                                            14D   
Shift                                                                               PM   
Date                                                                               NaN   
Hectare Squirrel Number                                                            NaN   
Age                                                                              Adult   
Primary Fur Color                                                                 Gray   
Combination of Primary and Highlight Color                                       Gray+   
Location                                                                  Ground Plane   
Above Ground Sighter Measurement                                                 FALSE   
Running                                                                          False   
Chasing                                                                          False   
Climbing                                                                         False   
Eating                                                                           False   
Foraging                                                                         False   
Kuks                                                                             False   
Quaas                                                                            False   
Moans                                                                            False   
Tail flags                                                                       False   
Tail twitches                                                                    False   
Approaches                                                                       False   
Indifferent                                                                      False   
Runs from                                                                        False   
Lat/Long                                    POINT (-73.95

**Does something stand out to you?**

In the column `Combination of Primary and Highlight Color` for example, we have a value that is only `"+"`.
Again, there are multiple ways to deal with this. When you can't figure out what the value means (like in this case), you can drop it. Sometimes, when you are able to figure out which one is meant, you can replace it with the correct value. This is where the domain knowledge come into play, knowing which variable can be replaced with what, and what it could mean.

Another example, from another dataset containing vehicle types where the vehicle type was manually encoded.
You see how `"Tow truck"` appears in many different forms.
The right approach is to consolidate all of the different values of tow truck into a single one.

In [11]:
trucks = ["Pick-up Truck", "Tractor Truck Diesel", "Box Truck", "Tractor Truck Gasoline", "Tow Truck / Wrecker", "Beverage Truck", "truck", "Armored Truck", "TRUCK", "Truck", "Tow Truck", "FIRE TRUCK", "PICK-UP TRUCK", "BOX TRUCK", "Fire Truck", "box truck", "TRUCK FLAT", "FIRTRUCK", "TRUCK VAN", "FIRETRUCK", "FDNY TRUCK", "dump truck", "Fire truck", "UPS TRUCK", "fire truck", "FOOD TRUCK", "MAIL TRUCK", "USPS TRUCK", "tow truck", "TOW TRUCK", "Tow truck"]
clean_trucks = ["Tow Truck" if "tow" in truck.lower() else truck for truck in trucks]
clean_trucks

['Pick-up Truck',
 'Tractor Truck Diesel',
 'Box Truck',
 'Tractor Truck Gasoline',
 'Tow Truck',
 'Beverage Truck',
 'truck',
 'Armored Truck',
 'TRUCK',
 'Truck',
 'Tow Truck',
 'FIRE TRUCK',
 'PICK-UP TRUCK',
 'BOX TRUCK',
 'Fire Truck',
 'box truck',
 'TRUCK FLAT',
 'FIRTRUCK',
 'TRUCK VAN',
 'FIRETRUCK',
 'FDNY TRUCK',
 'dump truck',
 'Fire truck',
 'UPS TRUCK',
 'fire truck',
 'FOOD TRUCK',
 'MAIL TRUCK',
 'USPS TRUCK',
 'Tow Truck',
 'Tow Truck',
 'Tow Truck']

### Duplicates

Search for duplicates and remove them



In [12]:
# There are no duplicates
df.duplicated().any()

False

If there had been, call `df.drop_duplicates(inplace=True)` to remove them.

Again, this is just a short introduction to data cleaning of structured data, there are many more steps that one could optionally take.

## References and reading material

[Data cleaning challenge - Kaggle](https://www.kaggle.com/rtatman/data-cleaning-challenge-handling-missing-values)  
[Preparing data for a machine learning model - Dataquest](https://www.dataquest.io/blog/machine-learning-preparing-data/)  
[Preliminary data cleaning - Kaggle](https://www.kaggle.com/quannguyen135/preliminary-data-cleaning-with-python)  
[Pandas data cleaning cheatsheet - Elite Data Science](https://elitedatascience.com/python-cheat-sheet)